### PR TITLE
[SPARK-45178][SS] Fallback to execute a single batch for Trigger.AvailableNow with unsupported sources rather than using wrapper

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to change the behavior when user runs streaming query with Trigger.AvailableNow, which query has any source which does not support Trigger.AvailableNow. Instead of using wrapper implementation, this PR proposes to fall back to execute a single batch (a.k.a Trigger.Once). 

This PR introduces a new flag `spark.sql.streaming.triggerAvailableNowWrapper.enabled` to retain the behavior for advanced and extreme users. The flag is marked as internal since it's really only for extreme users who are concerned about behavioral change.

Minor details would be following:

* This PR does not use Trigger.Once, hence users won't see deprecation warning for Trigger.Once. 
* This PR will provide a warning log to inform the source(s) which doesn't support Trigger.AvailableNow, so that users can indicate which source(s) is/are preventing them to enjoy benefits of Trigger.AvailableNow.

### Why are the changes needed?

We have observed a data duplication issue with 3rd party data source when it's used with Trigger.AvailableNow. The source didn't support Trigger.AvailableNow, and unfortunately is also not played well with wrapper implementation.

We care more about possible correctness issue than better coverage of Trigger.AvailableNow, hence want to stop using wrapper implementation by default. We also care about not breaking existing query, so fallback to single batch execution rather than failing the query.

### Does this PR introduce _any_ user-facing change?

Yes, this introduces a behavioral change for streaming query with Trigger.AvailableNow which contains any source not supporting Trigger.AvailableNow.

### How was this patch tested?

Modified UT.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @zsxwing @brkyvz @viirya @anishshri-db Mind taking a look? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -2180,6 +2180,17 @@ object SQLConf {
...
 
+  val STREAMING_TRIGGER_AVAILABLE_NOW_WRAPPER_ENABLED =
+    buildConf("spark.sql.streaming.triggerAvailableNowWrapper.enabled")
+      .internal()
+      .doc("Whether to use the wrapper implementation of Trigger.AvailableNow if the source " +
+        "does not support Trigger.AvailableNow. Enabling this allows the benefits of " +
+        "Trigger.AvailableNow with sources which don't support it, but some sources " +
+        "may show unexpected behavior including duplication, data loss, etc. So use with " +
+        "extreme care! The ideal direction is to persuade developers of source(s) to " +
+        "support Trigger.AvailableNow.")
```
Comment (anishshri-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2191]: Is there any way to expedite this ? Any way to force developers to upgrade their sources to be compatible with latest Spark requirements ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -2180,6 +2180,17 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val STREAMING_TRIGGER_AVAILABLE_NOW_WRAPPER_ENABLED =
+    buildConf("spark.sql.streaming.triggerAvailableNowWrapper.enabled")
+      .internal()
+      .doc("Whether to use the wrapper implementation of Trigger.AvailableNow if the source " +
+        "does not support Trigger.AvailableNow. Enabling this allows the benefits of " +
+        "Trigger.AvailableNow with sources which don't support it, but some sources " +
+        "may show unexpected behavior including duplication, data loss, etc. So use with " +
```
Comment (anishshri-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2189]: When would this cause data loss ? Given there are so many dangers, is it worthwhile to expose this if we cannot reason about correctness ? Can we limit the use only in certain situations perhaps to prevent users from shooting themselves in the foot ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala [hunk 3]
```diff
@@ -28,6 +28,12 @@ import org.apache.spark.sql.connector.read.streaming
 class AvailableNowDataStreamWrapper(val delegate: SparkDataStream)
   extends SparkDataStream with SupportsTriggerAvailableNow with Logging {
 
+  // See SPARK-45178 for more details.
+  logWarning(s"Activating the wrapper implementation of Trigger.AvailableNow for source " +
+    s"[$delegate]. Note that this might introduce possibility of deduplication, dataloss, " +
+    s"correctness issue. Enable the config with extreme care. We strongly recommend to contact " +
+    "with data source developer to support Trigger.AvailableNow.")
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala:35]: Nit: `to contact the Spark data source developer to support`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala [hunk 4]
```diff
@@ -201,7 +207,15 @@ case class MemoryStream[A : Encoder](
 
   override def initialOffset: OffsetV2 = LongOffset(-1)
 
+  override def prepareForTriggerAvailableNow(): Unit = synchronized {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala:210]: This is not strictly related to this change right ? basically we are trying to add support for AvailNow for memory source here ? Should we move that to a separate PR ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala [hunk 5]
```diff
@@ -52,11 +52,40 @@ class MicroBatchExecution(
...
+  protected def getTrigger(): TriggerExecutor = {
+    assert(sources.nonEmpty, "sources should have been retrieved from the plan!")
+    trigger match {
+      case t: ProcessingTimeTrigger => ProcessingTimeExecutor(t, triggerClock)
+      case OneTimeTrigger => SingleBatchExecutor()
+      case AvailableNowTrigger =>
+        // See SPARK-45178 for more details.
+        if (sparkSession.sqlContext.conf.getConf(
+            SQLConf.STREAMING_TRIGGER_AVAILABLE_NOW_WRAPPER_ENABLED)) {
+          logInfo("Configured to use the wrapper of Trigger.AvailableNow.")
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:66]: Can we log some more query specific info here ? or we don't have more info here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala [hunk 6]
```diff
@@ -52,11 +52,40 @@ class MicroBatchExecution(
...
+          MultiBatchExecutor()
+        } else {
+          val supportsTriggerAvailableNow = sources.distinct.forall { src =>
+            val supports = src.isInstanceOf[SupportsTriggerAvailableNow]
+            if (!supports) {
+              logWarning(s"source [$src] does not support Trigger.AvailableNow. Failing back to " +
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:72]: Nit: Typo ? `Falling back to` ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2191]: I'd say we are going to break a bunch of 3rd party data sources then, which is what we have been avoided so far.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2189]: This new config is marked as "internal", which is not going to be a part of public documentation. We expect very advanced users and/or operators only know about the existence of the config and use it as their own risk. We don't even expose this config in the warning message - that's the way I avoid users from shooting themselves on the foot.

Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2189]: I'm enumerating up the possibility - what we actually observed is a duplication, but suppose the data source which does not rely on offset management from Spark and tries to maintain that from source itself, then additional call might end up skipping some data to be provided.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala [hunk 7]
```diff
@@ -52,11 +52,40 @@ class MicroBatchExecution(
...
+  @volatile protected[sql] var triggerExecutor: TriggerExecutor = _
+
+  protected def getTrigger(): TriggerExecutor = {
+    assert(sources.nonEmpty, "sources should have been retrieved from the plan!")
+    trigger match {
+      case t: ProcessingTimeTrigger => ProcessingTimeExecutor(t, triggerClock)
+      case OneTimeTrigger => SingleBatchExecutor()
+      case AvailableNowTrigger =>
+        // See SPARK-45178 for more details.
+        if (sparkSession.sqlContext.conf.getConf(
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:64]: Not sure Im reading this correctly. If the flag is enabled, we use multi-batch directly ? Shouldn't this be the case if the flag is disabled ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala (hunk 3 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala:35]: maybe I'd remove Spark here - they need to contact 3rd party data source developer, not Spark community.

Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala:35]: ok sounds good

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala (hunk 4 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala:210]: While I agree this is a side-effect, this heavily increases a test coverage as well. We test with three data sources in TriggerAvailableNowSuite, and none of three supports Trigger.AvailableNow, hence we are basically missing the case where data source supports Trigger.AvailableNow in that suite. We missed that for a long time, I'd like to handle this as well, as long as we are here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala (hunk 7 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:64]: The logic is correct, but I agree that it's very confusing. I'll add some comment to elaborate.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala (hunk 5 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:66]: Yeah let's put queryID and runID at least to differentiate the query. Please note that this is an INFO log indicating that they turned on the flag (which is OK), and the WARN log will follow when they are falling back to use wrapper (which could be problematic).

**Review by anishshri-db — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala [hunk 8]
```diff
@@ -201,7 +207,15 @@ case class MemoryStream[A : Encoder](
...
+    availableNowEndOffset = latestOffset(initialOffset, ReadLimit.allAvailable())
+  }
+
   override def latestOffset(): OffsetV2 = {
+    throw new IllegalStateException("Should not reach here!")
+  }
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala:216]: Not sure why this cannot be called. Cannot `MemoryStream` be used as a `MicroBatchStream` after this change?

File: docs/ss-migration-guide.md [hunk 9]
```diff
@@ -26,6 +26,10 @@ Note that this migration guide describes the items specific to Structured Stream
 Many items of SQL migration can be applied when migrating Structured Streaming to higher versions.
 Please refer [Migration Guide: SQL, Datasets and DataFrame](sql-migration-guide.html).
 
+## Upgrading from Structured Streaming 3.5 to 4.0
+
+- Since Spark 4.0, Spark falls back to single batch execution if any source in the query does not support `Trigger.AvailableNow`. This is to avoid any possible correctness, duplication, and dataloss issue due to incompatibility between source and wrapper implementation. (See [SPARK-45178](https://issues.apache.org/jira/browse/SPARK-45178) for more details.)
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply docs/ss-migration-guide.md:31]: Let's discuss about this after the fix is landed in master branch. It is technically a behavioral change, hence easier to introduce the change in new major/minor version but may need some discussion to introduce the change in bugfix version. Maybe I can send a mail in dev@ and try to gather voices on it.

[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply docs/ss-migration-guide.md:31]: I intentionally avoid saying it is Trigger.Once. We deprecated it in good reason, and I'd say it is still worth saying users have to use Trigger.AvailableNow. We just have a fallback to Trigger.Once in technical reason, unfortunately.

Ideally, we still need to persuade 3rd party to implement Trigger.AvailableNow, but I also see that several data source projects having no update for a couple of years, which is unfortunate. Maybe we shouldn't introduce fallback logic and don't support the source so that 3rd party would indicate the necessity. My bad.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala (hunk 8 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala:216]: When data source implements AdmissionControl, this method is guaranteed to be never called. latestOffset(start, limit) will supersede this.

**Review by viirya — COMMENTED** [review]
As we are already in the status that sources not supporting `Trigger.AvailableNow` can work with the wrapper instead of simply failing since `Trigger.Once` is deprecated, it sounds reasonable to continue making it work under single batch execution.

One option is to stop support of it in 4.0 (as it is deprecated since 3.4), but add fallback logic for previous branches which provide the support for now (as it is related to correctness)?

**Comment by HeartSaVioR** [comment]
Looks like python side env issue. I'll rebase and see the chance to be already fixed in latest master.

**Review by viirya — APPROVED** [review]
I'm okay to continue the support with a fallback mode behind the scene. We probably can remove the support in future releases.

**Comment by HeartSaVioR** [comment]
> As we are already in the status that sources not supporting Trigger.AvailableNow can work with the wrapper instead of simply failing since Trigger.Once is deprecated, it sounds reasonable to continue making it work under single batch execution.
>
> One option is to stop support of it in 4.0 (as it is deprecated since 3.4), but add fallback logic for previous branches which provide the support for now (as it is related to correctness)?

Removing API would trigger a long discussion and debate. From my experience of Spark community (you've experienced longer than me :) ), I don't believe the community wants to remove the API unless there is a clear proof that few users use it. I think that's probably Spark 5.0, not 4.0.

That said, we couldn't apply more aggressive alternatives. I don't feel like we can persuade community to decide existing query to fail, which users cannot make any change to mitigate (it would require 3rd party data source engineer to work on).

**Comment by HeartSaVioR** [comment]
Ah I commented too late :) Thanks for your support!

**Comment by HeartSaVioR** [comment]
https://github.com/HeartSaVioR/spark/runs/16947355981
Looks like the test is flaky rather than consistent failure, it passed before rebasing and also passed locally.

https://github.com/HeartSaVioR/spark/actions/runs/6234669032
This was the CI before rebasing.

**Comment by HeartSaVioR** [comment]
Thanks, merging to master.
(I'll probably raise a discussion to dev@ for porting back to 3.5/3.4.)

**Comment by HeartSaVioR** [comment]
https://lists.apache.org/thread/ljronxf6bymvqjmlwpzy84gzgvnqrmoh
^^^ DISCUSSION thread in dev@.
